File: maven/WindnTrees0.0.1/application/src/main/java/app/controllers/world/Country.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package app.controllers.world;

import javax.validation.Valid;
import windntrees.CRUDController;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.Errors;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import windntrees.data.ResultObject;
import data.world.City;
import windntrees.CRUDInterface;
import windntrees.CRUDSource;
import data.world.CityRepository;

/**
 *
 * @author shams
 */
@Controller
@RequestMapping(value = {"/world/city", "/world/city/index"})
public class Country extends CRUDController<City> {
    
    @Autowired
    CityRepository cityRepository;
    
    @Override
    protected CRUDInterface<City> getRepository() {
        return (CRUDInterface<City>) new CRUDSource<City> (cityRepository) ;
    }

    @Override
    protected City generateNewKey(City object) {
        return object;
    }
    
    @RequestMapping(method = RequestMethod.GET)
    public String index(Model model)
    {
        return "world/country";
    }

    @Override
    protected City postCreate(City object) {
        if (object.getCountry() != null) {
            object.getCountry().setCityCapital(null);
        }
        return object; //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    protected City postUpdate(City object) {
        if (object.getCountry() != null) {
            object.getCountry().setCityCapital(null);
        }
        return object; //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    protected City postRead(City object) {
        if (object.getCountry() != null) {
            object.getCountry().setCityCapital(null);
        }
        return object; //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    protected Page<City> postRead(Page<City> page) {
        
        for (int i = 0; i < page.getContent().size(); i++) {
            if (page.getContent().get(i).getCountry() != null) {
                page.getContent().get(i).getCountry().setCityCapital(null);
            }

            if (page.getContent().get(i).getCountryCapital() != null) {
                page.getContent().get(i).getCountryCapital().setCityCapital(null);
            }
        }
        
        return page; //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    protected Page<City> postSelect(Page<City> page) {
        
        for (int i = 0; i < page.getContent().size(); i++) {

            if (page.getContent().get(i).getCountry() != null) {
                page.getContent().get(i).getCountry().setCityCapital(null);
            }

            if (page.getContent().get(i).getCountryCapital() != null) {
                page.getContent().get(i).getCountryCapital().setCityCapital(null);
            }

        }
        
        return page; //To change body of generated methods, choose Tools | Templates.
    }
    
    @Override
    public HttpEntity<ResultObject> create(@Valid @RequestBody City objectForm, 
            Errors errors) {
        
        if(objectForm.getCountryCapital() != null) {
            //if country capital reference is set then clear existing country city capitals
            cityRepository.clearCapital(objectForm.getCountryCapital().getCode());
        }
        
        return super.create(objectForm, errors); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public HttpEntity<ResultObject> update(@Valid @RequestBody City objectForm, 
            Errors errors) {
        
        if(objectForm.getCountryCapital() != null) {
            //if country capital reference is set then clear existing country city capitals
            cityRepository.clearCapital(objectForm.getCountryCapital().getCode());
        }
        
        return super.update(objectForm, errors); //To change body of generated methods, choose Tools | Templates.
    }
    
}

File: maven/WindnTrees0.0.1/application/src/main/webapp/resources/contents/world/Country.js
/* 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

function Country(data) {
    var instance = this;
    
    instance._datakey = data.code;
    
    instance.code = ko.observable(data.code);
    instance.code.extend({required: true, maxLength: 3});
    instance.code2 = ko.observable(data.code2);
    instance.code2.extend({required: true, maxLength: 2});
    
    instance.name = ko.observable(data.name);
    instance.name.extend({required: true, maxLength: 52});
    instance.localName = ko.observable(data.localName);
    instance.localName.extend({required: true, maxLength: 45});
    instance.governmentForm = ko.observable(data.governmentForm);
    instance.governmentForm.extend({required: true, maxLength: 45});
    instance.headOfState = ko.observable(data.headOfState);
    instance.headOfState.extend({maxLength: 60});
    
    instance.continent = ko.observable(data.continent);
    instance.continent.extend({required: true, maxLength: 13});
    instance.region = ko.observable(data.region);
    instance.region.extend({required: true, maxLength: 26});
    
    instance.surfaceArea = ko.observable(data.surfaceArea);
    instance.surfaceArea.extend({required: true, maxLength: 10});
    instance.population = ko.observable(data.population);
    instance.population.extend({required: true, maxLength: 11});
    
    instance.indepYear = ko.observable(data.indepYear);
    instance.indepYear.extend({maxLength: 4});
    
    instance.lifeExpectancy = ko.observable(data.lifeExpectancy);
    instance.lifeExpectancy.extend({maxLength: 5});
    
    instance.gnp = ko.observable(data.gnp);
    instance.gnp.extend({maxLength: 10});
    
    instance.gnpold = ko.observable(data.gnpold);
    instance.gnpold.extend({maxLength: 10});
    
    instance.capital = ko.observable(data.capital);
    instance.capital.extend({maxLength: 11});
    
    instance.cityCapital = ko.observable(data.cityCapital);
}
